A crucial aspect of working effectively with lists in Python is understanding how Python applies its internal indexing
structure to lists. For this assignment you will be creating, searching, and sorting lists you will create from the contents
of a small text file containing data values that are separated by commas (also commonly referred to as a ‘CSV’ file
(Comma Separated Values)) using some of the concepts we are learning about this week.
Start by downloading the M3_Data40.txt file to your local environment.
Then, open a new Jupyter Notebook and copy in the following small Python code snippet:

In [3]:
# Importing the csv package
import csv

# Setting up path for local file and storing in variable
with open('C:/Users/srikr/Downloads/M3_Data40.txt') as csvfile:
    readCSV = csv.reader(csvfile)
    for row in readCSV:
        print(row)
        
# Above code is for importing, given data from local file to read file in python environment

['Diagnosis', 'Age', 'Gender', 'PainType', 'BloodPres', 'Chol']
['FALSE', '63', 'MALE', 'TYPICAL_ANGINA', '145', '233']
['TRUE', '67', 'MALE', 'ASYMPTOMATIC', '160', '286']
['TRUE', '67', 'MALE', 'ASYMPTOMATIC', '120', '229']
['FALSE', '37', 'MALE', 'NON_ANGINAL', '130', '250']
['FALSE', '41', 'FEMALE', 'ATYPICAL_ANGINA', '130', '204']
['FALSE', '56', 'MALE', 'ATYPICAL_ANGINA', '120', '236']
['TRUE', '62', 'FEMALE', 'ASYMPTOMATIC', '140', '268']
['FALSE', '57', 'FEMALE', 'ASYMPTOMATIC', '120', '354']
['TRUE', '63', 'MALE', 'ASYMPTOMATIC', '130', '254']
['TRUE', '53', 'MALE', 'ASYMPTOMATIC', '140', '203']
['FALSE', '57', 'MALE', 'ASYMPTOMATIC', '140', '192']
['FALSE', '56', 'FEMALE', 'ATYPICAL_ANGINA', '140', '294']
['TRUE', '56', 'MALE', 'NON_ANGINAL', '130', '256']
['FALSE', '44', 'MALE', 'ATYPICAL_ANGINA', '120', '263']
['FALSE', '52', 'MALE', 'NON_ANGINAL', '172', '199']
['FALSE', '57', 'MALE', 'NON_ANGINAL', '150', '168']
['TRUE', '48', 'MALE', 'ATYPICAL_ANGINA', '110', '229']
['FA

This code will read the contents of a file containing data values that are separated by commas in a line-by-line manner
and print the contents of each line in the form of a Python list. Each item within that list will contain the corresponding
element of the comma-separated line within the data file from which it came.
Once you've run the code snippet, you will see that each line of the file contains six distinct values. For example, a print
of the result of reading the first line will show the following:
['FALSE', '63', 'MALE', 'TYPICAL_ANGINA', '145', '233']
These six distinct values represent attributes of hospital patients who had complained of chest pain. Specifically we have
the following:
- Diagnosis (i.e., was the patient diagnosed with heart disease?)
- Age
- Gender
- Type of chest pain
- Blood pressure reading
- Cholesterol level
Now that you are familiar with the content of the file, complete the following 6 tasks

Your first task is to read the file again and extract these six attributes from each line of the file and create
six distinct Python list objects (one for each of the six attributes) comprised solely of the values you extracted for a
given attribute. In other words, you should have separate lists for diagnosis, age, gender, etc. For example, the first
five "chest pain type" values should be as follows:
['TYPICAL_ANGINA', 'ASYMPTOMATIC', 'ASYMPTOMATIC', 'NON_ANGINAL', 'ATYPICAL_ANGINA']
One way to complete this task would be to use the code snippet from above as a starting point: you could replace
the “print(row)” statement with whatever Python code you feel is necessary to create the required list objects.
When finished, each of your list objects should contain 40 data values.
Note that the lists you have created thus far are composed solely of character strings. However, three of the
attributes (Age, Blood Pressure, and Cholesterol) are clearly meant to be used as integer data values. As such, before
proceeding any further, convert the contents of each of the ‘Age’, ‘Blood Pressure’, and ‘Cholesterol’ lists to integer
value

In [4]:
# Initialize empty lists for each attributes
diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol = [], [], [], [], [], []

# Reading the CSV file from Downloads and skip the header row as to exclude data impurity
with open('C:/Users/srikr/Downloads/M3_Data40.txt') as csvfile:
    readCSV = csv.reader(csvfile)
    next(readCSV)  # Skip the header row

    # Use list comprehensions to extract data and populate lists
    data = [row for row in readCSV]
    print(data)
    diagnosis = [row[0] for row in data]
    age = [int(row[1]) for row in data]
    gender = [row[2] for row in data]
    chest_pain_type = [row[3] for row in data]
    blood_pressure = [int(row[4]) for row in data]
    cholesterol = [int(row[5]) for row in data]
    
# data is know lists of list. as we know sequence of list like diagnosis, age, gender, chest_pain_type, blood_pressure and cholesterol with help of indexing and int casting technique we can segregate them as per requirement.
# List comprehension offers a shorter syntax, to create a new list based on the values of an existing list and printing the values for each attributes
print(diagnosis)
print(age)
print(gender)
print(chest_pain_type)
print(blood_pressure)
print(cholesterol)

[['FALSE', '63', 'MALE', 'TYPICAL_ANGINA', '145', '233'], ['TRUE', '67', 'MALE', 'ASYMPTOMATIC', '160', '286'], ['TRUE', '67', 'MALE', 'ASYMPTOMATIC', '120', '229'], ['FALSE', '37', 'MALE', 'NON_ANGINAL', '130', '250'], ['FALSE', '41', 'FEMALE', 'ATYPICAL_ANGINA', '130', '204'], ['FALSE', '56', 'MALE', 'ATYPICAL_ANGINA', '120', '236'], ['TRUE', '62', 'FEMALE', 'ASYMPTOMATIC', '140', '268'], ['FALSE', '57', 'FEMALE', 'ASYMPTOMATIC', '120', '354'], ['TRUE', '63', 'MALE', 'ASYMPTOMATIC', '130', '254'], ['TRUE', '53', 'MALE', 'ASYMPTOMATIC', '140', '203'], ['FALSE', '57', 'MALE', 'ASYMPTOMATIC', '140', '192'], ['FALSE', '56', 'FEMALE', 'ATYPICAL_ANGINA', '140', '294'], ['TRUE', '56', 'MALE', 'NON_ANGINAL', '130', '256'], ['FALSE', '44', 'MALE', 'ATYPICAL_ANGINA', '120', '263'], ['FALSE', '52', 'MALE', 'NON_ANGINAL', '172', '199'], ['FALSE', '57', 'MALE', 'NON_ANGINAL', '150', '168'], ['TRUE', '48', 'MALE', 'ATYPICAL_ANGINA', '110', '229'], ['FALSE', '54', 'MALE', 'ASYMPTOMATIC', '140', '23

Your second task entails using the updated Cholesterol list to find the list index values of each hospital
patient having a “cholesterol” reading that is greater than 250. Create a new list object with your result. HINT: you
can accomplish this task by searching the list of cholesterol values you created as part of the first task. Be sure to
print your results.

In [6]:
# Above we have cholesterol list, with help of enumerate we can get the index and it's values of patients with cholesterol > 250
high_cholesterol_index_patients = [i for i, value in enumerate(cholesterol) if value > 250]

# Print the result 
print("List Index Values of Patients with Cholesterol > 250:", high_cholesterol_index_patients)

List Index Values of Patients with Cholesterol > 250: [1, 6, 7, 8, 11, 12, 13, 18, 19, 21, 22, 26, 32, 37, 38]


Your third task is to find the "gender" value for each hospital patient having a "cholesterol" value that is
greater than 250. Create a new list to store your findings and be sure to print your results.

In [7]:
# Already we had high_cholesterol_index_of_patients, as index of these high_cholesterol_index_patients pass that index in gender lists and Find the "gender" value for patients with cholesterol > 250
gender_high_cholesterol = [gender[i] for i in high_cholesterol_index_patients]

# Print the results
print("Gender of patients with cholesterol > 250:", gender_high_cholesterol)

Gender of patients with cholesterol > 250: ['MALE', 'FEMALE', 'FEMALE', 'MALE', 'FEMALE', 'MALE', 'MALE', 'FEMALE', 'MALE', 'FEMALE', 'MALE', 'FEMALE', 'MALE', 'MALE', 'MALE']


Your fourth task is to find the index value for each hospital patient having an “age” value that is greater
than or equal to 50 and a “pain type” value that is not "ATYPICAL_ANGINA". Create a new list to store your findings
and be sure to print your results.

In [8]:
# Find index values for patients with age >= 50 and non-'ATYPICAL_ANGINA' pain type
selected_patients_indices = [i for i in range(len(age)) if age[i] >= 50 and chest_pain_type[i] != 'ATYPICAL_ANGINA']

# Print the results
print("Index values of selected patients:", selected_patients_indices)

Index values of selected patients: [0, 1, 2, 6, 7, 8, 9, 10, 12, 14, 15, 17, 20, 21, 23, 24, 25, 26, 27, 30, 31, 32, 33, 37, 38, 39]


Class Objects

As we have learned, Python’s class objects allow us to manage related data in an “object oriented” fashion. For the
remaining tasks in this assignment you will be constructing and applying your own Python class object to the
M3_Data40.txt data.

Create a new Python class to store and manage the information contained within the M3_Data40.txt
file. Your new class should be named “PatientData” and should include a class variable ‘PatientCount’ that
represents a counter indicative of the number of PatientData objects that have been instantiated within your
current Python session. The class must also contain seven instance variables including a unique ID for each patient
(use the name ‘uniqueID’ for this instance variable) and six additional instance variables corresponding to the six
different hospital patient attributes contained within the M3_Data40.txt file. Each time you create a new instance of
the class the ‘uniqueID’ instance variable value for the new object should be assigned the existing value of the
“PatientCount” class variable. The new instance should then increment the ‘PatientCount’ class variable by 1 (after
assigning its previous value to the ‘uniqueID’ instance variable as mentioned above.

In [9]:
class PatientData:
    # Class variable to count instances
    PatientCount = 0

    def __init__(self, diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol):
        # Increment the class variable and assign it to uniqueID
        PatientData.PatientCount += 1
        self.uniqueID = PatientData.PatientCount

        # Assign other instance variables
        self.diagnosis = diagnosis
        self.age = age
        self.gender = gender
        self.chest_pain_type = chest_pain_type
        self.blood_pressure = blood_pressure
        self.cholesterol = cholesterol

Read the M3_Data40.txt file again and store its content into a list of PatientData class objects, with
each row within the file being loaded into a distinct PatientData class object within that list. If you have defined your
PatientData class correctly according to the specifications provided in Problem 5, each class object within your list
should then contain the content of one row of the data file, with each hospital patient attribute value having been
assigned to a distinct instance variable within the class object, as well as a distinct uniqueID value that was
automatically assigned by the PatientData class when you read the corresponding row from the data file into the
class object. Then, using your list of PatientData objects, find the uniqueID for each hospital patient having an ‘age’
value of less than 56 and a “diagnosis” value of "TRUE". Create a new list to store your findings and be sure to print
your results.

In [10]:
# Reading Data and Creating PatientData Objects
patient_data_list = []

# Read the CSV file again and skip the header row
with open('C:/Users/srikr/Downloads/M3_Data40.txt') as csvfile:
    readCSV = csv.reader(csvfile)
    next(readCSV)  # Skip the header row

    # Create PatientData objects and append to the list
    for row in readCSV:
        diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol = row # the values from the CSV row converting into variables.
        age = int(age) # typecasting to integer
        blood_pressure = int(blood_pressure) # typecasting to integer
        cholesterol = int(cholesterol) # typecasting to integer
        patient = PatientData(diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol)
        patient_data_list.append(patient)

#  Access the age and diagnosis through variable patient and Find Unique IDs for Patients having age less than 56 and diagnosis as TRUE
selected_patients = [patient.uniqueID for patient in patient_data_list if patient.age < 56 and patient.diagnosis == 'TRUE']

# Print the results
print("Unique IDs for patients with age < 56 and diagnosis == 'TRUE':", selected_patients)

Unique IDs for patients with age < 56 and diagnosis == 'TRUE': [10, 17, 30, 37, 39]


Be sure to include some commentary within formatted Markdown cells explaining your approach to solving each of
the individual problems. When you are finished, save the Jupyter Notebook containing your work and commentary and
upload / submit it within the provided M3 Assignment Canvas submission portal. Be sure to save your Notebook using
the following nomenclature: first initial_last name_M3_assn" (e.g., J_Smith_M3_assn_).

In [11]:
import csv

# Task 1: Reading the CSV File and Creating Lists
diagnosis = []
age = []
gender = []
chest_pain_type = []
blood_pressure = []
cholesterol = []

# Read the CSV file and skip the header row
with open('C:/Users/srikr/Downloads/M3_Data40.txt') as csvfile:
    readCSV = csv.reader(csvfile)
    next(readCSV)  # Skip the header row

    # Extract data and populate lists
    for row in readCSV:
        diagnosis.append(row[0])
        age.append(int(row[1]))
        gender.append(row[2])
        chest_pain_type.append(row[3])
        blood_pressure.append(int(row[4]))
        cholesterol.append(int(row[5]))

# Task 2: Finding Cholesterol Values > 250
high_cholesterol_index_patients = [i for i, chol in enumerate(cholesterol) if chol > 250]

# Task 3: Finding Gender for Patients with High Cholesterol
gender_high_cholesterol = [gender[i] for i in high_cholesterol_index_patients]

# Task 4: Finding Index Values for Patients with Age >= 50 and Non-ATYPICAL_ANGINA Pain Type
selected_indices = [i for i, (a, pain) in enumerate(zip(age, chest_pain_type)) if a >= 50 and pain != 'ATYPICAL_ANGINA']

# Task 5: Creating the PatientData Class
class PatientData:
    PatientCount = 0
    
    def __init__(self, diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol):
        PatientData.PatientCount += 1
        self.uniqueID = PatientData.PatientCount
        self.diagnosis = diagnosis
        self.age = age
        self.gender = gender
        self.chest_pain_type = chest_pain_type
        self.blood_pressure = blood_pressure
        self.cholesterol = cholesterol

# Task 6: Reading Data and Creating PatientData Objects
patient_data_list = []

# Read the CSV file again and skip the header row
with open('C:/Users/srikr/Downloads/M3_Data40.txt') as csvfile:
    readCSV = csv.reader(csvfile)
    next(readCSV)  # Skip the header row

    # Create PatientData objects and append to the list
    for row in readCSV:
        diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol = row
        age = int(age)
        blood_pressure = int(blood_pressure)
        cholesterol = int(cholesterol)
        patient = PatientData(diagnosis, age, gender, chest_pain_type, blood_pressure, cholesterol)
        patient_data_list.append(patient)

# Task 6 (Continued): Finding Unique IDs for Patients Meeting Criteria
selected_patients = [patient.uniqueID for patient in patient_data_list if patient.age < 56 and patient.diagnosis == 'TRUE']

# Print the results for all tasks
print("Task 2: List Index Values of Patients with Cholesterol > 250:", high_cholesterol_index_patients)
print("Task 3: Gender of Patients with High Cholesterol:", gender_high_cholesterol)
print("Task 4: Index Values for Patients with Age >= 50 and Non-ATYPICAL_ANGINA Pain Type:", selected_indices)
print("Task 6: Unique IDs for Patients with Age < 56 and Diagnosis == 'TRUE':", selected_patients)

Task 2: List Index Values of Patients with Cholesterol > 250: [1, 6, 7, 8, 11, 12, 13, 18, 19, 21, 22, 26, 32, 37, 38]
Task 3: Gender of Patients with High Cholesterol: ['MALE', 'FEMALE', 'FEMALE', 'MALE', 'FEMALE', 'MALE', 'MALE', 'FEMALE', 'MALE', 'FEMALE', 'MALE', 'FEMALE', 'MALE', 'MALE', 'MALE']
Task 4: Index Values for Patients with Age >= 50 and Non-ATYPICAL_ANGINA Pain Type: [0, 1, 2, 6, 7, 8, 9, 10, 12, 14, 15, 17, 20, 21, 23, 24, 25, 26, 27, 30, 31, 32, 33, 37, 38, 39]
Task 6: Unique IDs for Patients with Age < 56 and Diagnosis == 'TRUE': [10, 17, 30, 37, 39]


Write a loop that accepts a string and produces a new string with the characters in reverse order. For example: 'analytics' would become 'scitylana'. Make use of Python's __input__ function to capute the string from a user.

In [12]:
# Accept a string from the user
input_string = input("Enter a string: ")
print("Input string:", input_string)

# Initialize an empty string to store the reversed characters
reversed_string = ""

# Iterate through the characters in the input_string in reverse order
for char in reversed(input_string):
    reversed_string += char

# Print the reversed string
print("Reversed string:", reversed_string)

Input string: analytics
Reversed string: scitylana


Write a loop that accepts a string and produces a list composed of items that each have at least 2 sequential characters from the original string, with the exception of the last element of the list which will contain any remaining individual character from the original string. For example: 'analytics' would become ['an', 'al', 'yt', 'ic', 's']

In [13]:
# Accept a string from the user
input_string = input("Enter a string: ")
print("Input string:", input_string)

# Initialize an empty list to store the items
items_list = []

# Initialize a variable to store the current item
current_item = ""

# Iterate through the characters of the input string
for char in input_string:
    # Add the current character to the current item
    current_item += char

    # Check if the current item has at least two characters
    if len(current_item) >= 3:
        items_list.append(current_item[:-1])  # Append all but the last character
        current_item = char  # Start a new current item with the current character

# Append the last current item to the list
items_list.append(current_item)

# Print the list of items
print("List of items:", items_list)

Input string: analytics
List of items: ['an', 'al', 'yt', 'ic', 's']
